cmd/opentelemetry-prometheus-sidecar/main.go

```diff
@@ -237,52 +296,28 @@ func Main() bool {
 	level.Debug(logger).Log("msg", "entering run state")
 	healthChecker.SetRunning()
 
-	// Run two inter-depdendent components:
-	// (1) Prometheus reader
-	// (2) Queue manager
-	// TODO: Replace this with x/sync/errgroup
-	var g run.Group
-	{
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting Prometheus reader")
-				err = prometheusReader.Run(ctx, startOffset)
-				return err
-			},
-			func(err error) {
-				// Prometheus reader needs to be stopped before closing the TSDB
-				// so that it doesn't try to write samples to a closed storage.
-				// See the use of `stopCh` below to explain how this works.
-				level.Info(logger).Log("msg", "stopping Prometheus reader")
-				cancelMain()
-			},
-		)
-	}
-	{
-		stopCh := make(chan struct{})
-		g.Add(
-			func() error {
-				if err := queueManager.Start(); err != nil {
-					return err
+	g := getRunGroup(ctx, cancelMain, logger, startOffset, prometheusReader, queueManager)
+
+	for {
+		if err := g.Run(); err != nil {
+			// TODO: don't use a string compare here
```

Comment:
:joy: I knew this would happen! I created an `ErrRestartReader` error there in tail/tail.go that I was thinking would be a signal to restart. It's being returned in a similar scenario where the sidecar is waiting for a WAL segment that is missing. We could modify the "truncated WAL segment" message to return ErrRestartReader and then use that error string here, potentially.

Reply:
I like this idea, then we could in theory restart the reader without restarting the sidecar completely for all these cases